nixos/modules/config/system-path.nix

```diff
@@ -7,6 +7,118 @@ with lib;
 
 let
 
+  # This type can accept three different types of expressions:
+  #   * A list of packages like `[ pkgs.foo pkgs.bar ]`
+  #   * An attribute set of packages like `{ inherit (pkgs) foo bar; }`
+  #   * A more structured attribute set of packages like `{ "foo".package = pkgs.foo.override { hasBar = false; }; }`
+  #
+  # The first two cases will be coerced into the third one, which allows the
+  # package to be disabled and its package and list of outputs to be overridden.
+  packagesType =
+    let
+      inherit (lib) types;
+
+      outputsFor = drv:
+        if drv.outputSpecified or false
+        then [ drv.outputName ]
+        else ((drv.meta or {}).outputsToInstall or [ (drv.outputName or "out") ]);
```

Comment:
The same logic is in `pkgs.buildEnv`: https://github.com/NixOS/nixpkgs/blob/670182de729294664672addedf6ecef2095037c5/pkgs/build-support/buildenv/default.nix#L60-L70

Would be good to deduplicate this, probably into its own function in `lib.derivations.*`